Upgrade and move javascript dependencies in `static/third` to npm

refeed opened the pull request:
### Upgrade and move javascript dependencies in `static/third` to npm

I've tested all of these changes pretty thoroughly. If you find any error on this, please tell us.

**NOTE**
> Always remember to bump up the `PROVISION_VERSION` variable in `version.py` for each of these commits (just for dependency upgrade commits)

Upgraded:
- [x] `winchan.js` Upgraded to `0.2.0` and moved to npm. **(MERGED)**

- [x] `String.prototype.codepointat` Upgraded to `0.2.0` and moved to npm. **(MERGED)**

- [x] `Underscore.js` Upgraded to `1.8.3` and moved to npm. **(MERGED)**
**Notes**: Has minified version in `node_modules` directory. Should we use it ?

- [x] `xdate` Moved to npm (has no updates in npm). **(MERGED)**
**Notes**: Has a minor version update (`0.8.1`) in GitHub, not released in npm yet.

- [x] `jquery-mousewheel` Upgraded to `3.1.6` and moved to npm.
**Notes**: Not the latest version (3.1.13) because Zulip has an issue on version >= 3.1.7. The issue is on the `static/js/message_list.js:638`, we can't handle `mousewheel` event there.
- [x] `jquery-autosize` Upgraded to `3.0.21` and moved to npm.
**Notes**: Has minified version in `node_modules` directory. Should we use it ?


- [x] `perfect-scrollbar` Upgraded to `0.7.1` and moved to npm.  **(MERGED)**
**Notes**: Has minified version in `node_modules` directory. Should we use it ?

- [x] `handlebars` Upgraded to `3.0.3` and moved to npm.  **(MERGED)**
**Notes**: Not the latest version because we have some issue with `handlebars` version >= 4.0.0. This is the release notes for handlebars https://github.com/wycats/handlebars.js/blob/master/release-notes.md

Not upgraded yet:
- [ ] `jquery-form`
**Notes**: Cannot upgrade because there is an issue with newer version, when I click on "Get API key button" the API responses `{"msg":"CSRF Error: CSRF token missing or incorrect.","result":"error"}`.

- [ ] `jquery-idle`
**Notes**: The software has been modified.

- [ ] `html5-formdata`
**Notes**: There is no updates for this dependency, and has no npm package (https://github.com/francois2metz/html5-formdata )

- [ ] `jquery.caret`
**Notes**: Has no npm package.

- [ ] `jquery-filedrop`
**Notes**: Has no npm package, and the software has been modified.

- [ ] `jquery-throttle-debounce`
**Notes**: Has no npm package and any updates.

- [ ] `sorttable`
**Notes**: Has no npm package and any updates since 2007.

- [ ] `marked.js`
**Notes**: The sofware has been modified, and probably we can't submit the patch to the upstream, because the changes is about Zulip.

- [ ] `lazyload`
**Notes**: Not maintained anymore and has no npm package.

- [ ] `spectrum`
**Notes**: The software has been modified.

- [ ] `bootstrap-notify.js`
**Notes**: The software has been modified.

- [ ] `bootstrap.js`
**Notes**: The software has been modified.

Fixes #1709

smarx commented:
*Automated message from Dropbox CLA bot*

@rafidaslam, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
@rafidaslam this looks great!  I'm going to merge the first few of these commits.

One note, though: please only put the `Fixes: #1709` in the final commit in this sort of series; I'm going to have to remove that from the commits I merge today to avoid closing the issue when I merge the first one.

timabbott commented:
OK, a few notes on these:
* We don't need to use the minified versions of these third-party modules, though it's possible we'll later want to start doing so using the new feature Tommy added to our javascript build process.  
* For jquery-autosize, we were using an event that they used to listen for back when they were jQuery only: `$("#new_message_content").trigger("autosize.resize");`.  I bet we just replace that with calling `autosize($("#new_message_content"))`, but worth testing.

timabbott commented:
I merged the first few commits.

timabbott commented:
@rafidaslam I think it's probably reasonable for these to not patch version.py, and I'll just do a version.py update when I merge them (just worried about unnecessary merge conflicts while rebasing).  

Huge thanks for working on this!  I'm going to hold off on autosize, handlebars and perfectscrollbar until after the 1.5 release just for safety, but feel free to keep working on the rest in this PR!

rafidaslam commented:
Oh, ok, thanks for the review ! I'll keep working on this, also thanks for the autosize feedback

rafidaslam commented:
@timabbot updated autosize, I think for `$("#new_message_content").trigger("autosize.resize");` it's should be `autosize.update($("#new_message_content"))`, since when I test fill a long message to the compose box, and click the send button, it's resizing, but when I test with `autosize()` it's not resizing.

timabbott commented:
I just merged the handlebars commit.  Thanks @refeed!  

I know we're still debugging the mousewheel/scrollbar issues; is autosize ready to review?

timabbott commented:
And it was very quick to upgrade handlebars the rest of the way; yay :)

timabbott commented:
For autosize, you should have added it to `.eslintrc.json`, not added the global comment.

timabbott commented:
OK, I tried testing the autosize change, and it seems to not work as expected; in particular, if you close the compose box, and reopen it, the size is the same as it was before (even if now it's empty and before it needed to be 5 lines tall).  Maybe the `autosize.update` function doesn't work like that?

timabbott commented:
I think the perfectscrollbar issue is this:
```
subprocess.check_call(['./manage.py', 'collectstatic', '--no-default-ignore',                                                        
                       '--noinput', '-i', 'assets', '-i' 'node_modules'],                                                            
                      stdout=fp, stderr=fp)                                                                                          
```

`mange.py collectstatic` does 2 things: it copies static asset files to `prod-static/` and then it does a post-processing step (with our custom code in `zerver/storage.py`).  In general, we don't want `node_modules` in production, since it is huge, so we pass `-i node_modules`  above to exclude it from what is copied.  But I guess the `post_process` state for CSS files needs the originals to be present in the copied locatio for some reason.

You should be able to reproduce locally using `tools/build-release-tarball test`, and then then error will be at the end of the log file of the form `/tmp/tmp.t8O9aICS4B/update-prod-static.log`.

It's probably a significant code-reading and/or print statement adding exercise to figure out how best to fix the issue "right"; there's probably a workaround that involves having the script copying that CSS file to somewhere else.

I think a plausibly acceptable workaround would be to add a new bit of code that runs just after `tools/minify-js` that just copies the relevant CSS file from `node_modules` to the path expected as shown in the error message.

timabbott commented:
@refeed let me know if you need any help with figuring that out :)

refeed commented:
@timabbott I've fixed the compose autosizing by moving `clear_box()` in `compose.cancel()` to the [top](https://github.com/zulip/zulip/pull/3227/files#diff-3aa67e805d80237a4d4064d2f2fb74a6R303).

And about `perfectscrollbar`, I've fixed it by hardcoding the `update-prod-static` to manually copy the `perfectscrollbar` css to `STATIC_ROOT`, I'm not sure is this fine?

timabbott commented:
@refeed I think #3244 and #4071 added some merge conflicts; can you rebase?  

Your CSS copying code is fine; I'm OK with a hacky solution for copying from node_modules to production for now.  Though we don't depend on `rsync` currently so I think you should create the directory a bit more manually.

refeed commented:
@timabbott rebased, and I've changed `rsync` with `cp`

timabbott commented:
Sigh, I was sick when you updated this, and now there are more merge conflicts.  @refeed do you by any chance have time to rebase this again?  It'd be much appreciated!!

refeed commented:
@timabbot sorry for late reply, rebased, and updated `perfectScrollbar` and `autosize` to the latest version :smile:

tejaskasetty commented on zproject/settings.py:
```diff
@@ -730,6 +730,7 @@ def get_secret(key):
                 'third/thirdparty-fonts.css',
                 'third/jquery-perfect-scrollbar/css/perfect-scrollbar.css',
```
@refeed This file won't exist. And shouldn't this be removed?

refeed replied:
thanks!! sure that should be removed

timabbott commented:
@refeed would you be up for rebasing this, now that the perfect-scrollbar piece is merged?  Thanks!

refeed commented:
@timabbott rebased :wink:

timabbott commented:
This is likely still a useful reference for upgrading these packages, but unlikely to be directly rebased.  Closing, thanks for all your work @refeed!

tommyip commented:
For the ones that we have patched, we can move the library to npm and then apply our changes to it with https://github.com/ds300/patch-package. I tried this with spectrum a while ago and it worked fine, though patch-package only support one patch file per library, so things like bootstrap which have a lot of modifications might become messy.

timabbott commented:
@tommyip hmm, interesting.  I think that's potentially an option, but before trying that we should first see if upstream is active; if so, it'd be better to submit patches upstream and run off a fork with those patches.  

If you look at `git log --follow -p static/third/spectrum/spectrum.js`, almost all of our changes are things upstream almost certainly did (0ea770fc180e498ef45ca49dec88f8edd8de868a and eda1cdd0c524009c897adacacda69ab2fa7c3241 being the potential exceptions).

andersk commented:
For spectrum, the current upstream version seems to work for us unmodified: #12749.
